File: backend/src/main/java/com/sistemacontable/service/UsuarioService.java
package com.sistemacontable.service;

import com.sistemacontable.dto.RolDto;
import com.sistemacontable.dto.UsuarioRequest;
import com.sistemacontable.dto.UsuarioResponse;
import com.sistemacontable.entity.Role;
import com.sistemacontable.entity.Usuario;
import com.sistemacontable.repository.RoleRepository;
import com.sistemacontable.repository.UsuarioRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Servicio para gestionar operaciones de usuarios
 */
@Service
@Transactional
public class UsuarioService {
    
    @Autowired
    private UsuarioRepository usuarioRepository;
    
    @Autowired
    private RoleRepository roleRepository;
    
    /**
     * Obtiene todos los usuarios del sistema
     * @return lista de usuarios como DTOs
     */
    public List<UsuarioResponse> obtenerTodosLosUsuarios() {
        return usuarioRepository.findAll().stream()
            .map(usuario -> new UsuarioResponse(
                usuario.getId(),
                usuario.getNombreUsuario(),
                usuario.getRol().getNombre()
            ))
            .collect(Collectors.toList());
    }
    
    /**
     * Crea un nuevo usuario en el sistema
     * @param usuarioRequest datos del usuario a crear
     * @return usuario creado como DTO
     * @throws RuntimeException si el username ya existe o el rol no existe
     */
    public UsuarioResponse crearUsuario(UsuarioRequest usuarioRequest) {
        // Verificar que el nombre de usuario no exista
        if (usuarioRepository.existsByNombreUsuario(usuarioRequest.getNombreUsuario())) {
            throw new RuntimeException("El nombre de usuario ya existe");
        }
        
        // Buscar el rol
        Role rol = roleRepository.findById(usuarioRequest.getRolId())
            .orElseThrow(() -> new RuntimeException("Rol no encontrado"));
        
        // Crear usuario con contraseña hasheada
        Usuario usuario = new Usuario(
            usuarioRequest.getNombreUsuario(),
            usuarioRequest.getContrasena(), // Sin encriptar para simplicidad
            rol
        );
        
        Usuario usuarioGuardado = usuarioRepository.save(usuario);
        
        return new UsuarioResponse(
            usuarioGuardado.getId(),
            usuarioGuardado.getNombreUsuario(),
            usuarioGuardado.getRol().getNombre()
        );
    }
    
    /**
     * Elimina un usuario del sistema
     * @param usuarioId ID del usuario a eliminar
     * @throws RuntimeException si el usuario no existe
     */
    public void eliminarUsuario(Integer usuarioId) {
        if (!usuarioRepository.existsById(usuarioId)) {
            throw new RuntimeException("Usuario no encontrado");
        }
        
        usuarioRepository.deleteById(usuarioId);
    }
    
    /**
     * Obtiene todos los roles disponibles
     * @return lista de roles como DTOs
     */
    public List<RolDto> obtenerTodosLosRoles() {
        return roleRepository.findAll().stream()
            .map(rol -> new RolDto(rol.getId(), rol.getNombre()))
            .collect(Collectors.toList());
    }
}

File: frontend/src/services/usuarioService.ts
import { UsuarioRequest, UsuarioResponse, RolDto } from '../types/auth';

const API_BASE_URL = 'http://localhost:8080/api';

/**
 * Servicio para manejar operaciones de gestión de usuarios
 */
class UsuarioService {

  /**
   * Obtiene headers con token de autorización
   * @returns headers con Authorization bearer token
   */
  private getAuthHeaders(): Record<string, string> {
    const token = localStorage.getItem('token');
    return {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`,
    };
  }

  /**
   * Obtiene todos los usuarios del sistema
   * @returns promesa con lista de usuarios
   */
  async obtenerUsuarios(): Promise<UsuarioResponse[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/usuarios`, {
        method: 'GET',
        headers: this.getAuthHeaders(),
      });

      if (!response.ok) {
        throw new Error('Error al obtener usuarios');
      }

      return await response.json();
    } catch (error) {
      console.error('Error al obtener usuarios:', error);
      throw error;
    }
  }

  /**
   * Crea un nuevo usuario
   * @param usuarioData datos del usuario a crear
   * @returns promesa con usuario creado
   */
  async crearUsuario(usuarioData: UsuarioRequest): Promise<UsuarioResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/usuarios`, {
        method: 'POST',
        headers: this.getAuthHeaders(),
        body: JSON.stringify(usuarioData),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(errorText || 'Error al crear usuario');
      }

      return await response.json();
    } catch (error) {
      console.error('Error al crear usuario:', error);
      throw error;
    }
  }

  /**
   * Elimina un usuario del sistema
   * @param usuarioId ID del usuario a eliminar
   * @returns promesa void
   */
  async eliminarUsuario(usuarioId: number): Promise<void> {
    try {
      const response = await fetch(`${API_BASE_URL}/usuarios/${usuarioId}`, {
        method: 'DELETE',
        headers: this.getAuthHeaders(),
      });

      if (!response.ok) {
        throw new Error('Error al eliminar usuario');
      }
    } catch (error) {
      console.error('Error al eliminar usuario:', error);
      throw error;
    }
  }

  /**
   * Obtiene todos los roles disponibles
   * @returns promesa con lista de roles como DTOs
   */
  async obtenerRoles(): Promise<RolDto[]> {
    try {
      const response = await fetch(`${API_BASE_URL}/usuarios/roles`, {
        method: 'GET',
        headers: this.getAuthHeaders(),
      });

      if (!response.ok) {
        throw new Error('Error al obtener roles');
      }

      return await response.json();
    } catch (error) {
      console.error('Error al obtener roles:', error);
      throw error;
    }
  }
}

export const usuarioService = new UsuarioService();